### DISCUSSION TITLE
SA search

### DISCUSSION THREAD

**Discussion by khromenokroman** [strongswan/strongswan#1868] [Q&A] [answered]
guys, tell me if I understood correctly that SA is stored in the kernel, if so, how can the function be seen ...
and the second question, tell me if I use kernel-libipsec, then SA will also be stored in the kernel or somewhere in user space
sorry for the stupid questions :)

**Comment by tobiasbrunner** [comment, marked as answer]
Maybe read [the introduction](https://docs.strongswan.org/docs/5.9/howtos/introduction.html) and [kernel-libipsec](https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html) docs.

**Reply by khromenokroman** [reply]
Thanks for the answer, I read everything

**Reply by khromenokroman** [reply]
but I still didn't understand, if I use kernel-libipsec, where will SA be, in the kernel?

**Reply by tobiasbrunner** [reply]
Are you sure you read it? The first thing the second doc states is (highlighting by me):

> The kernel-libipsec plugin for libcharon provides an IPsec backend that works **entirely in userland**...

And it even clarifies it in the second paragraph:

> Both other kernel interfaces, kernel-netlink (the default) and kernel-pfkey, install IPsec SAs in the operating system’s IPsec stack. This plugin provides an alternative, for instance, if the OS implementation does not support a required algorithm (e.g. AES-GCM on macOS).

**Reply by khromenokroman** [reply]
sorry for the stupid questions, I thought it meant only data plane, thank you very much for the answer

**Reply by khromenokroman** [reply]
tell me this question, can I separate the control plane and data plane, that is, I take strongswan in the standard ike sa core delivery, and I redirect the payload that is encrypted with esp to my module, and decrypt with those functions that are in kernel-libipsec and sa I take via netlink from the core, will I be able to do this? a hybrid version in which sa is in the core, and decryption in esp in the user space
Sorry if my questions seem silly, I'm just new to this issue.

**Reply by tobiasbrunner** [reply]
> decrypt with those functions that are in kernel-libipsec and sa I take via netlink from the core

I don't understand what you mean by that. Or what purpose that would have.

**Reply by khromenokroman** [reply]
I have my own solution, and I would like to integrate strongswan there, but I would like ike sa to be in the core and esp to be in userspace.
that is, I accept ike sa in the kernel, and I redirect the esp packet to myself in the decryptor that sa requests via netlink in the kernel and I use it to unpack the esp packet in userspace

**Reply by tobiasbrunner** [reply]
> I have my own solution

What does that mean?  Solution for what?

> I would like ike sa to be in the core and esp to be in userspace.

What does that mean? What is "the core"? The kernel? Why should IKE SAs that are managed by strongSwan in userland be in the kernel? And why would that be relevant for ESP SAs?

**Reply by khromenokroman** [reply]
What does that mean? Solution for what?

> ngfw

What does that mean? What is "the core"? The kernel? Why should IKE SAs that are managed by strongSwan in userland be in the kernel? And why would that be relevant for ESP SAs?

> I want to divide strongswan into 2 parts.
1 part which is control plane(IKE SA) negotiation and creation of ESP SA (which is placed in the kernel)
Part 2 is to pick up the package that should come to strongswan encrypted and decrypt it in your plugin (userspace), that is, strongswan will never receive the encrypted package, it will only be used to produce KE_SA_INIT, IKE_AUTH and CREATE_CHILD_SA

**Reply by tobiasbrunner** [reply]
> I want to divide strongswan into 2 parts.
> 1 part which is control plane(IKE SA) negotiation and creation of ESP SA (which is placed in the kernel)

That's already the default.

**Reply by khromenokroman** [reply]
yes, but part 2 is not, so I want to make a hybrid, in general, I have already figured it out, thank you :)
